torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,863 @@
+        return node.run(new_inputs)
+
+    def new_graph_id(self) -> GraphID:
+        return GraphID(next(self.graph_counter))
+
+    def new_func_id(self) -> FunctionID:
+        return FunctionID(next(self.func_counter))
+
+    def add_function(self, model, inputs, static_input_idxs) -> Callable:
+        id = self.new_func_id()
+        self.ids_to_funcs[id] = WrappedFunction(
+            model, remove_unaligned_input_idxs(inputs, static_input_idxs), id
+        )
+        fn = functools.partial(self.run, function_id=id)
+
+        # container needs to set clean up when fn dies
+        get_container().add_strong_reference(fn)
+
+        return fn
+
+    def get_roots(self) -> Generator[CUDAGraphNode]:
+        for nodes in self.roots.values():
+            for node in nodes:
+                yield node
+
+    @staticmethod
+    def get_curr_generation() -> int:
+        return GenerationTracker.generation
+
+    def try_end_curr_recording(self) -> None:
+        """
+        Check if the current recording can be terminated, either because all outputs of the
+        previously recorded node are dead or because it was executed in a different
+        generation. Will set current_node to None and in_recording to False if successful.
+        """
+        assert self.in_recording
+
+        if self.current_node is None:
+            self.in_recording = False
+            return
+
+        # multiple invocations, allow overwriting the previous generation
+        if self.current_gen != self.get_curr_generation():
+            # we need to keep track of the previous live outputs if they were in recording
+            # in case we need to record again
+            self.previous_recording_outputs.extend(
+                self.current_node.path_live_weakrefs()
+            )
+
+            self.clear_current_node_outputs_and_set_to_none()
+            self.in_recording = False
+            return
+
+        if self.current_node.all_outputs_are_dead():
+            self.clear_current_node_outputs_and_set_to_none()
+            self.in_recording = False
+            return
+
+    def try_end_curr_execution(self) -> None:
+        """
+        Check if the current executng node can be terminated, either because all outputs of the
+        previously executed node are dead or because it was executed in a different generation.
+        Will set current_node to None if successful.
+        """
+
+        assert not self.in_recording
+        if self.current_node is None:
+            return
+
+        if self.current_gen != self.get_curr_generation():
+            self.clear_current_node_outputs_and_set_to_none()
+            return
+
+        if self.current_node.all_outputs_are_dead():
+            self.clear_current_node_outputs_and_set_to_none()
+
+    def clear_current_node_outputs_and_set_to_none(self):
+        self.current_node.clear_path_outputs()
+        self.current_node = None
+
+    def convert_execution_to_recording(self):
```

Comment:
So there's some invariant this has to be called... after `record_function`, I guess??

Reply:
Actually, I think I misread this function name: this is taking us from "execution" mode and putting us in "recording" mode, right? "convert" is very misleading as I think of some sort of data structure conversion in that case.

Reply:
This function really can use the comment.

Reply:
Added a comment and renamed. let me know if you think it needs further clarification